src/libopensc/card-atrust-acos.c

```diff
@@ -720,8 +720,8 @@ static int atrust_acos_compute_signature(struct sc_card *card,
 				/* XXX: assume no hash is wanted */
 				flags = SC_ALGORITHM_RSA_HASH_NONE;
 			tmp_len = sizeof(sbuf);
-			r = sc_pkcs1_encode(card->ctx, flags, data, datalen,
-					sbuf, &tmp_len, sizeof(sbuf)*8, NULL);
+			r = sc_pkcs1_encode(card->ctx, flags, data, datalen, sbuf, &tmp_len, sizeof(sbuf) * 8,
+			                    NULL);
```

Comment:
Thats the intention to make it aligned regardless the tab length.

But you are right that this is not much used in the code. I think what you describe woudl be written as
```
AlignAfterOpenBracket: DontAlign
ContinuationIndentWidth: 2
UseTab: ForContinuationAndIndentation
```
but I am afraid it will make much longer lines for the `if` conditions. But let me try.

Reply:
I am playing with this a bit with the following configuration:
```
+TabWidth: 4
+IndentWidth: 4
 ContinuationIndentWidth: 8
-UseTab: AlignWithSpaces
+AlignAfterOpenBracket: DontAlign
+UseTab: ForContinuationAndIndentation
```
And the hardcoded `ContinuationIndentWidth` has a result of all the structs and tables definition are indented with two tabs, which is something I am not very happy with. Also if we have a function call with many arguments in the `if` condition, the overflowing arguments are indented with one more tab, which can go too much right (again depending on the tab width of the editor). The same for broken srings in the function arguments.

One solution for this problem might be not allowing function calls in the `if` conditions, because this makes arguably hard to read and debug code, long lines and excess indentation. We have similar guides in libssh: https://gitlab.com/libssh/libssh-mirror/-/blob/master/CONTRIBUTING.md#make-use-of-helper-variables but again, this does not fit the "current use in opensc" -- just a suggestion.

What I found confusing in this formatting configuration:

 * The long lines sometimes result in the following reformat, which I find ugly and I did not find a way to prevent it:
```
-idprime_read_binary(sc_card_t *card, unsigned int offset, unsigned char *buf, size_t count,
-               unsigned long flags)
+idprime_read_binary(
+               sc_card_t *card, unsigned int offset, unsigned char *buf, size_t count, unsigned long flags)
```
 * The structure/tables definitions are indented with 2 tabs instead of one:
```
        const u8 *aids[] = {
-               (const u8 *)"\xa0\x00\x00\x02\x47\x10\x01",
+                       (const u8 *)"\xa0\x00\x00\x02\x47\x10\x01",
```
 * The strings are still aligned with some spaces instead of tabs for some reason:
```
                                parse_error(cur, "Template insane: file-path length should not be "
-                                               "less than 2 bytes");
+                                                "less than 2 bytes");
```
The most crazy stuff is avoided with the `ColumnLimit: 0`, but in some other cases, it still tries to indent with spaces and I have no idea why in some cases where long function calls with a lot of arguments is used.

I just pushed the updated commits. Would be glad for another look if you would be more happy with them then I am.

Reply:
@frankmorgner there are some unresolved issues of the current configuration. Additionally, with the current configuration, I see some issues with alignment of continuation of function calls in `if(` conditions, such as:
```
@@ -3500,9 +3500,9 @@ wrap_key(CK_SESSION_HANDLE session)
 		util_fatal("Secret key (to be wrapped) not found");
 
 	if (!find_object(session, CKO_PUBLIC_KEY, &hWrappingKey, opt_object_id_len ? opt_object_id : NULL,
-			opt_object_id_len, 0))
+			    opt_object_id_len, 0))
 		if (!find_object(session, CKO_SECRET_KEY, &hWrappingKey,
-				opt_object_id_len ? opt_object_id : NULL, opt_object_id_len, 0))
+				    opt_object_id_len ? opt_object_id : NULL, opt_object_id_len, 0))
 			util_fatal("Public/secret key (wrapping key) not found");
 
 	rv = p11->C_WrapKey(session, &mechanism, hWrappingKey, hkey, pWrappedKey, &pulWrappedKeyLen);
```
I find this totally non-intuitive, as the three (or 4) tabs are followed by 4 spaces (that probably account for the `if (` string?). This all with `UseTab: Always` (this is not going away even with `UseTab: ForContinuationAndIndentation`). FYI I am using clang 14 now.

Reply:
I don't know exactly why, but this will actually result in tabs instead of spaces:
```
ContinuationIndentWidth: 20
AlignAfterOpenBracket: DontAlign
```

Reply:
hmm, the above setting will then result in a mix of tabs and spaces when used in a case like this:
```

int i =         //  VeryVeryVeryVeryVeryLongComment
  longFunction( // Again a long comment
    arg);
```
So this is not yet solved

Reply:
In my tests, this here seems to do what you want:
```
TabWidth: 4
IndentWidth: 4
ContinuationIndentWidth: 8
AlignAfterOpenBracket: DontAlign
```

Reply:
Thanks for having a look. But with this configuration, I did not notice any significant change. It looks like it resolved the issue with the `if`, but there are very similar cases with `else if` case, which it still tries to format with tabs + spaces, see the example below:
```
@@ -1380,9 +1380,9 @@ int main(int argc, char * argv[])
 			find_object_flags(session, mf_flags, &object,
 					opt_object_id_len ? opt_object_id : NULL, opt_object_id_len, 0);
 		} else if (!find_object(session, CKO_PUBLIC_KEY, &object,
-					opt_object_id_len ? opt_object_id : NULL, opt_object_id_len, 0) &&
-				!find_object(session, CKO_CERTIFICATE, &object,
-					opt_object_id_len ? opt_object_id : NULL, opt_object_id_len, 0))
+						   opt_object_id_len ? opt_object_id : NULL, opt_object_id_len, 0) &&
+				   !find_object(session, CKO_CERTIFICATE, &object,
+						   opt_object_id_len ? opt_object_id : NULL, opt_object_id_len, 0))
 			util_fatal("Public key nor certificate not found");
 	}
 
```
I also still get all the structure initializers indented with two tabs which I find as a huge waste of horizontal space (see the results of the current CI run). But otherwise, I find it mostly usable.

One more thing to consider is alignment of `type variable comment` initializers. This is used quite a lot in the code, but not much in the new one as I do not find it very useful. Example:
```
 	unsigned char sk_enc[16];	/* encrypt session key */
 	unsigned char sk_mac[16];	/* mac session key */
 	unsigned char icv_mac[16];	/* instruction counter vector(for sm) */
-	unsigned char bFipsCertification;	/* fips mode Alg */
+	unsigned char bFipsCertification; /* fips mode Alg */
```
The formatter tries to get rid of the additional whitespace, causing large diffs too. In this case, this is aligned with tabs so it looks completely different in different editors. What would be your take on these changes?

Reply:
```
UseTab: AlignWithSpaces
```
fixes this:
>Use tabs for line continuation and indentation, and spaces for alignment

I think we can agree on the fact that clang-format will never provide a perfect result. And the above example of mixing tabs and spaces for indentation sounds like a bug in clang-format.

In my tests I've noticed that my preferred option `AlignAfterOpenBracket: AlwaysBreak` together with `ColumnLimit: 0` results in a removal of intentionally placed line breaks. To fix this, I *could* argue that we should not hold some changes back to reduce the number of changes in this pull request (i.e. the change to `ColumnLimit: 110`). But as said before, I think I can live with the imperfections we are currently seeing and I'd like to move on with the format in its current state.

Reply:
I reformated the structures to be indented with 2 tabs. Now, some of them are also formatted with `AlignArrayOfStructures: Left`, but this option keeps crashing my clang so I left it out yet.
There are still some weirdnesses (and the commit series is mess now) so I would propose to merge only the documentation + .clang-format + gh action (I will split it) and the rest of reformatting only after some better review. Its too large.

Reply:
I pused couple of more changes built with clang 15, where this option seems to be working. I did some changes that looked like made sense to me, but ignored some where we did not completely agreed or they were only reflowing ling breaks to longer lines.

With that I am still not completely confident the clang does the right thing in the following cases:
 * the `else if` continuation indentation is weird (as already mentioned)
 * multiline the ternary operator is still indented with spaces
 * with `AlignArrayOfStructures`, if there is many spaces used to align the structure members, some of the spaces are replaced with tabs, which breaks the indentation (at least in the diff view, which probably does not follow our tab width -- but if it looks different with different settings, its kind of useless)
 * aligning of type + variable + values + comments in the variable declaration -- half of the code does it and the other does not do that. What would be your take?

Reply:
As said before, it doesn't need to be perfect. For now, it is nice to
- have at least some definition written down and
- automate the process of highlighting formatting problems in PRs

😮‍💨 

Reply:
Yeah. Thats what I tried to put in contributing.md Please, check if it covers all what is needed.